### DISCUSSION TITLE
[Plugin Request] Python Graphene

### DISCUSSION THREAD

**Discussion by adrianschneider94** [dotansimha/graphql-code-generator#2516] [General]
Hi,

I'm currently writing a graphql-codegen-plugin that generates Python graphene (https://graphene-python.org) types from a schema.

What should I pay attention to if I want to contribute the plugin? Code style, patterns etc...

So far I extended the BaseTypesVisitor and get fine results for most schemas. Do you have a collection of schemas I can test against?

Kind regards,
Adrian Schneider

**Comment by dotansimha** [comment]
That's awesome, @adrianschneider94 ! 
Our plugins are written in TypeScript, and you can take a very simple boilerplate from plugins like `add` or `time`. 
You can either manage your plugin in a separate repo, or in this repo (this way we'll be able to help more and able to make sure we don't break the new plugin with changes). 

Extending `BaseTypesVisitor` is indeed the correct approach and should make it easier for you to write the rest of the plugin logic. 

In this repo there is a directory called `dev-test` with experiments and tests we are doing, you can use it easily (add your plugin to the repo, build it, add configuration to `dev-test/codegen.yml` and run with `yarn generate:examples`). 

You can also find a list of GraphQL apis to test here: https://github.com/APIs-guru/graphql-apis

Feel free to contact me or anyone from our team if you need any help!

**Comment by adrianschneider94** [comment]
Things, that don't work yet:

**Extending Interfaces** https://github.com/graphql-python/graphene/issues/1068
Interfaces itself work, but extending them doesn't. Python does not allow (in contrast to GraphQL, Javascript etc.) to use classes that appear later in the source file. To circumvent it, you can use lambda expressions for the fields (https://github.com/graphql-python/graphene/issues/110).
But this does not work with interfaces yet. Another approach would be to build a dependency tree and reorder the classes, but this would still not solve circular dependencies and would be quite complex.

**Directives** https://github.com/graphql-python/graphene/issues/1069
There is some support in the core library, but it's not really documented and I don't know yet, how to bring it together with Graphene.

**Comment by dotansimha** [comment]
@adrianschneider94 this is awesome! 
I think the codegen plugin you are writing should support the existing features of graphql-python so it sounds like a good process! 
I think sorting the output could work and solve this issue, or build a dep tree with libraries like https://www.npmjs.com/package/dependency-graph and then sort it automatically. 

Let me know how it goes and if we can help :)

**Comment by jeremy-deutsch** [comment]
I wrote something like this in TS a couple months ago - if you want, I can export the function that turns a schema AST into a python code string

https://github.com/jeremy-deutsch/graphene-codegen

**Comment by eladn** [comment]
Here is a `quicktype`s TS core script that emits python code from a json schema. This might be useful.

https://github.com/quicktype/quicktype/blob/master/src/quicktype-core/language/Python.ts

**Comment by moshere-fb** [comment]
> Hi,
> 
> I'm currently writing a graphql-codegen-plugin that generates Python graphene (https://graphene-python.org) types from a schema.
> 
> What should I pay attention to if I want to contribute the plugin? Code style, patterns etc...
> 
> So far I extended the BaseTypesVisitor and get fine results for most schemas. Do you have a collection of schemas I can test against?
> 
> Kind regards,
> Adrian Schneider

Hi Adrian,
Do you've any prototype available already?

Thank you

**Comment by adrianschneider94** [comment]
@moshere-fb 
https://gist.github.com/adrianschneider94/9ddb7f6c906fd13f7c215c3a10a9749c

**Comment by dotansimha** [comment]
@adrianschneider94 this is awesome!!! Do you want to create a PR for this plugin? :) We can help you with the implementation if you need.

**Comment by hyusetiawan** [comment]
@adrianschneider94 / @dotansimha  any update on this, would be super timely if this is available :)

**Comment by Andrew-Talley** [comment]
@adrianschneider94 Are you still working on this project? I'm about to wrap up a base python-types plugin (#4146), so I'd be happy to pick up where you left off and try to finish up this work.

**Comment by yanncks** [comment]
Is this still being worked on? Would love to have this

**Comment by adrianschneider94** [comment]
Hey guys, sorry, I haven‘t made any progress  since the last gist I posted. My use case vanished...

https://gist.github.com/adrianschneider94/9ddb7f6c906fd13f7c215c3a10a9749c

But the code worked quite well when I tried it the last time. Yet it‘s not properly tested and not cleaned up.

[Edit]
The issue with the interfaces can be resolved quite easily as graphene accepts strings that point to the referenced class instead of the class itself when you link fields.
